core/src/main/java/com/backblaze/b2/json/B2JsonBoundedByteArrayOutputStream.java

```diff
@@ -0,0 +1,185 @@
+        B2Preconditions.checkArgument(maxCapacity <= SYSTEM_MAX_CAPACITY, "maxCapacity must not be bigger than " + SYSTEM_MAX_CAPACITY);
+
+        this.maxCapacity = maxCapacity;
+        this.size = 0;
+
+        // initialize with the lesser of 64 and maxCapacity, expand later if needed
+        // keep initial capacity an internal detail (not exposed)
+        final int initialCapacity = Math.min(64, maxCapacity);
+        this.output = new byte[initialCapacity];
+    }
+
+    /**
+     * writes one byte into the output array
+     *
+     * @param i one byte
+     * @throws IOException if expanding capacity would cross
+     *                     the maxCapacity threshold
+     */
+    @Override
+    public synchronized void write(int i) throws IOException {
+        // check and expand capacity if needed, handle potential overflow in checkCapacity()
+        checkCapacity(1);
+
+        output[size] = (byte) i;
+        size++;
+    }
+
+    /**
+     * writes an array of bytes into the output array
+     *
+     * @param bytes input array of bytes
+     * @param offset offset for the input array
+     * @param length number of bytes to write
+     * @throws IOException if expanding capacity would cross
+     *                     the maxCapacity threshold
+     */
+    @Override
+    public synchronized void write(byte[] bytes, int offset, int length) throws IOException {
+        validateParameters(bytes, offset, length);
+
+        checkCapacity(length);
+
+        // append the new content to the output, and reset size
+        System.arraycopy(bytes, offset, output, size, length);
+        size += length;
+    }
+
+    /**
+     * makes a new String by decoding the bytes in output array
+     * using the specified charset.
+     *
+     * @param charsetName the charset to be used in decoding
+     * @return the string representation of output bytes after decoding
+     * @throws UnsupportedEncodingException if the named charset is not supported
+     */
+    public synchronized String toString(String charsetName) throws UnsupportedEncodingException {
+        return new String(output, 0, size, charsetName);
+    }
+
+    /**
+     * returns a copy of the output byte array
+     *
+     * @return a copy of output buffer
+     */
+    public synchronized byte[] toByteArray() {
+        // make a copy of internal output content
+        return Arrays.copyOf(output, size);
+    }
+
+    /**
+     * returns the max capacity of output array
+     *
+     * @return the max capacity
+     */
+    public int getMaxCapacity() {
+        return this.maxCapacity;
+    }
+
+    /**
+     * returns the current output array size
+     *
+     * @return output output array size
+     */
+    public int getSize() {
+        return this.size;
+    }
+
+    /**
+     * checks and expands capacity if needed
+     *
+     * @param length the new input byte length
```

Comment:
This could be clearer.  I think this parameter is the additional capacity needed, in addition to the current size (not the current capacity).

Reply:
Right, updated.